gcc/d/d-codegen.cc

```diff
@@ -1828,12 +1828,52 @@ d_mark_read (tree exp)
   return exp;
 }
 
-// Build equality expression between two RECORD_TYPES T1 and T2.
+// Return TRUE if the struct SD is suitable for comparison using memcmp.
+// This is because we don't guarantee that padding is zero-initialized for
+// a stack variable, so we can't use memcmp to compare struct values.
+
+bool
+identity_compare_p(StructDeclaration *sd)
+{
+  if (sd->isUnionDeclaration())
+    return true;
+
```

Comment:
I think it is a reasonable choice to say that all safety belts are off when comparing two unions.  You should really be comparing ``ba.a == bb.a`` to guarantee that the inner struct is compared correctly.

No, comparing ``Bar b1 == Bar b2`` will not succeed in your case.  Yes, this is different from dmd.  Should it be considered a bug/regression?  I don't mind reverting this for unions, but it might not always be  correct anyway...